keystone_json_assignment/json.py

```diff
@@ -76,56 +78,68 @@ def _get_public_id(self, user_id):
             'local_id': user_id,
             'entity_type': 'user'})
 
+    def load_user_project_map(self):
+        mtime = datetime.datetime.utcfromtimestamp(
+            os.path.getmtime('/etc/keystone/user-project-map.json'))
+        if self.userprojectmap == {} or mtime > self.last_loaded:
+            with open('/etc/keystone/user-project-map.json', 'r') as f:
+                userprojectmap = yaml.load(f)
+                self.last_loaded = mtime
+
+            projectnames = set()
+            for projects in userprojectmap.values():
+                for project in projects:
+                    projectnames.add(project)
+
+            projectidcache = {}
+            for projectname in projectnames:
+                try:
+                    project = self.resource_manager.get_project_by_name(
+                        projectname, CONF.identity.default_domain_id)
+                    projectidcache[projectname] = project['id']
+                except exception.ProjectNotFound as e:
+                    LOG.warning("keystone-json-assignment: "
+                                "project-map: cannot lookup %s: %s",
+                                project, e.message)
+
+            for user, projects in userprojectmap.items():
+                projectids = {}
+                projectid = None
+                user_id = self._get_public_id(user)
+                if user_id:
+                    self.useridmap[user_id] = user
+                for projectname in projects:
+                    projectid = projectidcache.get(projectname)
+                    if projectid:
+                        projectids[projectid] = 1
+                self.userprojectmap[user] = projectids
```

Comment:
Here we're just updating the existing dict. With other words removed users remain inside the dict, which is unwanted.

How about we start with an empty temporary local dict and at the end just replace the member attribute version with it? This way deleting users is also handled.